extensions/2.0/Khronos/KHR_skin_strict/README.md

```diff
@@ -0,0 +1,47 @@
+# KHR\_skin\_strict
+
+## Contributors
+
+* TBD
+
+## Status
+
+Draft
+
+## Dependencies
+
+Written against the glTF 2.0 spec.
+
+## Overview
+
+`KHR_skin_strict` introduces no new features or schema to the glTF file format. Instead, it defines a strict subset of the existing skinning specification, and imposes requirements — beyond those of the original specification — on tools creating glTF assets with a `skins` entry.
+
+Implementations of skinned mesh animation vary between engines, for historical and performance reasons. This stricter skinning definition increases portability and simplicity for client implementations, at the cost of some flexibility and increased responsibilities for authoring tools.
+
+When used, `KHR_skin_strict` must be an *optional* extension. Because it is a subset of the original specification, clients that do not recognize the extension can always safely read files that include the extension, so inclusion in the `extensionsRequired` list is neither necessary nor allowed. Presence of the `KHR_skin_strict` extension acts as a hint, for clients that recognize it, that the safer/stricter skinning requirements have been met.
+
+Client implementations may choose to require that glTF assets containing one or more skins include the `KHR_skin_strict` extension, and reject assets that do not. Authoring tools are advised to include the extension whenever possible.
+
+## Requirements
+
+* All skins *must* define the `skeleton` property.
+* The `skeleton` node for a skin *must* be a joint, and the direct parent of the entire joint subtree for that skin.
+* All joints in a skin *must* define a single fully-connected subtree, sharing a single root joint.
+* A non-joint node *must not* be both an ancestor and a descendant of a joint node.
+    * `node ⊃ joint ⊃ joint ⊃ node` is valid.
+    * `node ⊃ joint ⊃ node ⊃ joint` is invalid, regardless of whether the two joints are part of the same skin.
```

Comment:
The intent is to forbid "child skins"? Like this?

![child_skin](https://user-images.githubusercontent.com/11024420/72930389-7424b800-3d21-11ea-99a8-2b85a0556ed7.png)

Is this what reduz meant here

> A mesh *must* only be able to bind to a single skin. No game engine supports binding a mesh to multiple skins. If an exporter wants to do this, it needs to join both skeletons.

Reply:
I'm not sure if reduz meant the same thing, but it would be forbidden by this wording, yes. A weaker restriction could be:

> A non-joint node *must not* be both an ancestor and a descendant of a joint node, **if the two joints are in the same skin**.

I'm unable to think of a compelling reason for nesting one skeleton under another; it sounds unlikely to work consistently... any concerns with prohibiting it?

Reply:
To clarify:

1. The main intent of this line is, strongly, to prohibit non-joint nodes being injected into a the joint subtree of a single skin.
2. Prohibiting one skin from being nested under another is a side effect that (IMO) seems beneficial, but I don't feel as strongly about that part.

^Although, the requirement as-written would not prohibit nesting a skin under another skin _without_ an intervening node. If we intend to keep (2), that should also be prohibited for consistency.

Reply:
> A non-joint node must not be both an ancestor and a descendant of a joint node, **if the two joints are in the same skin**.

This is already forbidden above by requiring the joints form a tree, right?

> I'm unable to think of a compelling reason for nesting one skeleton under another

If the intent is to forbid one skeleton under another, this is too weak: it only forbids them when there is an intervening non-joint node. This is still allowed

![child_skin](https://user-images.githubusercontent.com/11024420/72936363-a7207900-3d2c-11ea-8d8e-ac1d563746fa.png)

Possible phrasing to forbid this: "Given two joint nodes from different skins, neither may be a descendant of the other".

> any concerns with prohibiting it?

Well, an exporter can just extend the "upper" skin into the "lower" skin and join them together like reduz says. Just trying to understand the reason for it.

Reply:
> This is already forbidden in the bullet point above by requiring the joints form a tree, right?

I don't think it's explicit enough in the earlier bullet, but probably there are better ways to phrase that.  How about removing this bullet and rephrasing the one above:

>  All joints in a skin *must* define a single fully-connected subtree, sharing a single root joint, **with no non-joint nodes coming between the joints of the tree**.

If we do want to prohibit nested skins, that can be a clearer separate bullet.

Reply:
Ok (just saw your second reply above). I think making it part of the previous bullet is a good idea.

Reply:
Both figures by @scurest  are fine, with the assumption that one mesh binds to each skin.

Almost all game engines have a "Bone Attachment" or "Socket" system to parenting things directly to a joint node.
In most cases you would skins[0] -> skeletons[0], skins[1] -> skeletons[1], and there would be a bone attachment from the root joint of skeletons[1] to the appropriate one in skeletons[0].

What we need to prevent is that there doesn't exist a case where meshs[0] tries to bind to skins[0] and skins[1] simultaneously, that is a no go.

Reply:
> What we need to prevent is that there doesn't exist a case where meshs[0] tries to bind to skins[0] and skins[1] simultaneously, that is a no go.

What does that mean exactly? This?

    "nodes": [
      {"mesh": 0, "skin": 0},
      {"mesh": 0, "skin": 1},
      ...
    ]

Reply:
> 
> 
> > What we need to prevent is that there doesn't exist a case where meshs[0] tries to bind to skins[0] and skins[1] simultaneously, that is a no go.
> 
> What does that mean exactly? This?
> 
> ```
> "nodes": [
>   {"mesh": 0, "skin": 0},
>   {"mesh": 0, "skin": 1},
>   ...
> ]
> ```

Correct

Reply:
What about

    {"mesh": 0},
    {"mesh": 0, "skin": 0}

Reply:
Removed the confusing bullet and clarified the previous one.

@marstaik I don't understand your comment about not re-binding different instances of a mesh to different skins. A "mesh" in glTF is not instantiated until it is mounted at a node, and meshes may be instantiated multiple times. Assuming both skins have the same number of bones, why couldn't different instances of a mesh use different skins (or no skin)?

Certainly we don't want a single _instance_ of a mesh bound to joints from different skins...

Reply:
Sorry, I confused myself there. I was tired and missed the fact that it was a node list, haha.

> Certainly we don't want a single instance of a mesh bound to joints from different skins...

Yes you are correct